File: examples/MovementSystem/MovementSystem.cs
using Client;
using Engine.Core;
using Movement.V1;

namespace Examples;

public class MovementSystem : SystemBase
{
    private readonly EntityQuery _q;
    private ulong _tickCount;

    public MovementSystem()
    {
        _q = NewQuery()
            .With(Query.ReadWrite<Position>())
            .With(Query.ReadOnly<Velocity>());
    }

    protected override Task OnUpdateAsync()
    {
        foreach (var entity in _q.Entities)
        {
            var pos = _q.Get<Position>(entity);
            var vel = _q.Get<Velocity>(entity);
            _q.Set(entity, new Position
            {
                X = pos.X + vel.X * DeltaTime,
                Y = pos.Y + vel.Y * DeltaTime,
                Z = pos.Z + vel.Z * DeltaTime
            });
        }

        _tickCount++;
        if (_tickCount % 20 == 0 && _q.Entities.Count > 0)
        {
            var e = _q.Entities[0];
            var p = _q.Get<Position>(e);
            Console.WriteLine($"[Movement] Tick {_tickCount} | Entity {e.Id} Position: ({p.X:F2}, {p.Y:F2}, {p.Z:F2})");
        }

        return Task.CompletedTask;
    }
}

File: examples/typescript/movement-system/src/movement-system.ts
import { Query, SystemBase, type EntityQuery } from "@ecs/client";
import { PositionSchema, VelocitySchema } from "@ecs/protos/movement/v1/movement_pb.js";

/**
 * The same system as `examples/MovementSystem`, in TypeScript.
 *
 * It declares the same component types, by the same protobuf full names, and the
 * coordinator cannot tell which language it is written in — which is the point.
 */
export class MovementSystem extends SystemBase {
  private readonly moving: EntityQuery;

  constructor() {
    super();
    this.moving = this.newQuery()
      .with(Query.write(PositionSchema))
      .with(Query.read(VelocitySchema));
  }

  protected override onUpdate(): void {
    for (const entity of this.moving.entities) {
      const position = this.moving.get(entity, PositionSchema);
      const velocity = this.moving.get(entity, VelocitySchema);

      this.moving.set(entity, PositionSchema, {
        $typeName: PositionSchema.typeName,
        x: position.x + velocity.x * this.deltaTime,
        y: position.y + velocity.y * this.deltaTime,
        z: position.z + velocity.z * this.deltaTime,
      });
    }

    if (this.tickId % 20n === 0n && this.moving.entities.length > 0) {
      const first = this.moving.entities[0]!;
      const position = this.moving.get(first, PositionSchema);
      console.log(
        `[Movement/ts] Tick ${this.tickId} | Entity ${first} Position: ` +
          `(${position.x.toFixed(2)}, ${position.y.toFixed(2)}, ${position.z.toFixed(2)})`,
      );
    }
  }
}
